src/log.h

```diff
@@ -185,35 +190,120 @@ class LogOutputBuffer : public ICombinedLogOutput {
 	Logger &m_logger;
 };
 
+#ifdef __ANDROID__
+class AndroidLogOutput : public ICombinedLogOutput {
+	public:
+		void logRaw(LogLevel lev, const std::string &line);
+};
+#endif
 
-extern StreamLogOutput stdout_output;
-extern StreamLogOutput stderr_output;
-extern std::ostream null_stream;
+class LogTarget {
+public:
+	// Must be thread-safe. These can be called from any thread.
+	virtual bool hasOutput() = 0;
+	virtual void log(const std::string &buf) = 0;
+};
 
-extern std::ostream *dout_con_ptr;
-extern std::ostream *derr_con_ptr;
-extern std::ostream *derr_server_ptr;
 
-extern Logger g_logger;
+class StreamProxy {
+public:
+	StreamProxy(std::ostream *os) : m_os(os) { }
+
+	template<typename T>
+	StreamProxy& operator<<(T&& arg) {
+		if (m_os) {
+			*m_os << std::forward<T>(arg);
+		}
+		return *this;
+	}
 
-// Writes directly to all LL_NONE log outputs for g_logger with no prefix.
-extern std::ostream rawstream;
+	StreamProxy& operator<<(std::ostream& (*manip)(std::ostream&)) {
+		if (m_os) {
+			*m_os << manip;
+		}
+		return *this;
+	}
 
-extern std::ostream errorstream;
-extern std::ostream warningstream;
-extern std::ostream actionstream;
-extern std::ostream infostream;
-extern std::ostream verbosestream;
-extern std::ostream dstream;
+private:
+	std::ostream *m_os;
+};
 
-#define TRACEDO(x) do {               \
-	if (g_logger.getTraceEnabled()) { \
-		x;                            \
-	}                                 \
-} while (0)
+class LogStream {
+public:
+	LogStream() = delete;
+	DISABLE_CLASS_COPY(LogStream);
+
+	LogStream(LogTarget &target) :
+		m_target(target),
+		m_buffer(std::bind(&LogStream::internalFlush, this, std::placeholders::_1)),
+		m_dummy_buffer(),
+		m_stream(&m_buffer),
+		m_dummy_stream(&m_dummy_buffer),
+		m_proxy(&m_stream),
+		m_dummy_proxy(nullptr) { }
+
+	template<typename T>
+	StreamProxy& operator<<(T&& arg) {
+		StreamProxy& sp = m_target.hasOutput() ? m_proxy : m_dummy_proxy;
```

Comment:
just make this a no-no when `!m_target.hasOutput()`?

Reply:
I'm not sure what you mean by this. Can you explain?

Reply:
Nevermind, I missed that this function has to return a `StreamProxy &`